.github/workflows/phpunit-tests.yml

```diff
@@ -0,0 +1,128 @@
+
+# Since Unit Tests are required to pass for each PR,
+# we cannot disable them for documentation-only changes.
+on:
+    pull_request:
+    push:
+        branches:
+            - trunk
+            - 'release/**'
+            - 'wp/**'
+
+# Cancels all previous workflow runs for pull requests that have not completed.
+concurrency:
+    # The concurrency group contains the workflow name and the branch name for pull requests
+    # or the commit hash for any other events.
+    group: ${{ github.workflow }}-${{ github.event_name == 'pull_request' && github.head_ref || github.sha }}
+    cancel-in-progress: true
+
+jobs:
+    test-php:
+        name: ${{ matrix.php }}${{ matrix.multisite && ' multisite' || '' }} on ${{ matrix.os }}
+        runs-on: ${{ matrix.os }}
+        timeout-minutes: 20
+        if: ${{ github.repository == 'WordPress/gutenberg' || github.event_name == 'pull_request' }}
+        strategy:
+            fail-fast: false
+            matrix:
+                php:
+                    [
+                        '5.6',
+                        '7.0',
+                        '7.1',
+                        '7.2',
+                        '7.3',
+                        '7.4',
+                        '8.0',
+                        '8.1',
+                        '8.2',
+                    ]
+                os: [ubuntu-latest]
+                multisite: [false, true]
+
+        env:
+            WP_ENV_PHP_VERSION: ${{ matrix.php }}
+
+        steps:
+            - uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # v3.1.0
+
+            - name: Set up Node.js
+              uses: actions/setup-node@8c91899e586c5b171469028077307d293428b516 # v3.5.1
+              with:
+                  node-version-file: '.nvmrc'
+                  cache: npm
+
+            ##
+            # This allows Composer dependencies to be installed using a single step.
+            #
+            # Since the tests are currently run within the Docker containers where the PHP version varies,
+            # the same PHP version needs to be configured for the action runner machine so that the correct
+            # dependency versions are installed and cached.
+            ##
+            - name: Set up PHP
+              uses: shivammathur/setup-php@1a18b2267f80291a81ca1d33e7c851fe09e7dfc4 # v2.22.0
+              with:
+                  php-version: '${{ matrix.php }}'
+                  coverage: none
+
+            # Ensure that Composer installs the correct versions of packages.
+            - name: Override PHP version in composer.json
+              run: composer config platform.php ${{ matrix.php }}
+
+            # The spatie/phpunit-watcher package is not compatible with PHP < 7.2.
+            # It must be removed before running the tests.
+            - name: Remove incompatible Composer packages
+              if: ${{ matrix.php < '7.2' }}
+              run: composer remove spatie/phpunit-watcher --dev --no-update
```

Comment:
if this breaks in supported PHP versions, why not remove the dependency altogether? I'm worried about the disconnect between requiring this package and having this line in the workflow YML. best case scenario is that when this dependency is no longer required, we'll still be running this step, which is benign, but also is going to confuse someone why it's there.

is there a way to have composer tell us which packages are not compatible? or should we file a bug report with `spatie/phpunit-watch` to have that project list a minimum version if one is missing?

Reply:
> if this breaks in supported PHP versions, why not remove the dependency altogether?

This is because `spatie/phpunit-watcher` is required by the `npm run test:php:watch` command.
Thus it cannot be simply removed.
I was going to create a new GB issue to discuss the possibility of replacing it with a similar package that supports PHP < 7.2 (if that substitute package exists).

> is there a way to have composer tell us which packages are not compatible?

Yes, composer already does that. It knows that this package is not compatible with PHP < 7.2 and refuses to install it if the current PHP version is lower than 7.2. So it should be removed before installing other packages (otherwise `composer install` fails).

> or should we file a bug report with spatie/phpunit-watch to have that project list a minimum version if one is missing?

No, it's not missing. https://github.com/spatie/phpunit-watcher/blob/master/composer.json#L19

Reply:
> Yes, composer already does that. It knows that this package is not compatible with PHP < 7.2 and refuses to install it if the current PHP version is lower than 7.2. So it should be removed before installing other packages (otherwise composer install fails).

What I meant was that maybe we can avoid hard-coding a specific dependency as we've done here. It looks like if we spit out the tree format of `composer show` we can get that information, and then process it with a small node script.

```php
<?php

require 'vendor/autoload.php';

use Composer\Semver\Semver;

$PHP_VERSION = empty( getenv( 'PHP_VERSION' ) ) ? phpversion() : getenv( 'PHP_VERSION' );
$tree = json_decode( file_get_contents( 'php://stdin' ) );

function works_with_php_version( $version, $package ) {
	foreach ( $package->requires as $dep ) {
		if ( 'php' === $dep->name && ! Semver::satisfies( $version, $dep->version ) ) {
			//echo "Bailing on {$package->name} because of {$dep->version}" . PHP_EOL;
			return false;
		}

		if ( ! empty( $dep->requires ) && false === works_with_php_version( $version, $dep ) ) {
			return false;
		}
	}

	return true;
}

$unsupported = array();
foreach ( $tree->installed as $package ) {
	if ( ! works_with_php_version( $PHP_VERSION, $package ) ) {
		$unsupported[] = $package->name;
	}
}

echo implode( ',', $unsupported );
```

Invoked as `composer show --tree --format=json | php php-supported.php` this will spit out a list of packages that have dependencies which aren't supported on the running PHP version, or the version set in `PHP_VERSION` if such an ENV is set.

But it also includes another package not listed, and to the best of my ability to double-check it, it's right according to the listed minimum versions.

```
spatie/phpunit-watcher,yoast/phpunit-polyfills
```

While maybe someone registered the wrong minimum version, or we simply haven't hit the part that breaks yet, this is why I think it'd be good at least to explore determine this through inspecting the package definitions. The list will undoubtedly change more rapidly than we come around to update it.

Reply:
Thanks for sharing the snippet.
> we can avoid hard-coding a specific dependency as we've done here

Out of all Composer packages, this is the only package that gives trouble.
I didn't see the need to write a script for detecting such packages when I was working on the CI job, as this was the only "problematic" script.

`yoast/phpunit-polyfills` - I'm not sure I understand why it isn't compatible with PHP 5.6.
Check out this Packagist page:
https://packagist.org/packages/yoast/phpunit-polyfills
![Screenshot 2022-12-20 at 11 58 26](https://user-images.githubusercontent.com/43744263/208651277-e0cc5136-d06d-46d9-ae66-e2caf4af969c.png)

Also, if someone adds a composer package that is not compatible with PHP 5.6 or another Core-supported PHP version, I'd prefer that the author of that PR would be aware of that instead of removing or hiding such packages with the help of a script.

Reply:
> why it isn't compatible with PHP 5.6.

the dependencies of its dependencies state a minimum required version above 5.6. like I wrote above though, we may simply have not run into the places where it will break, or the package authors may have incorrectly listed their requirements.

if you run `composer show --tree` you will see where it comes in. ironically it's through `phpunit` so maybe we could safely eliminate that dependency chain as a given since we're already running.

> If someone adds a composer package that is not compatible with PHP 5.6 or another Core-supported PHP version, I'd prefer that the author of that PR would be able to see the error (i.e., composer install should fail) instead of removing or hiding such packages with the help of a script.

I totally agree! that's why I asked about removing the package. do we know if anyone would even miss the `npm run test:php:watch` scripts? I wouldn't, because I don't use that due to the way it obscures the underlying test runner and won't let me do what I want to do with `phpunit` (and it doesn't even run when I try, immediately fails with an error).

it's not the worst thing to have this hard-coded exception in the code, but I'd love to do all we can to avoid it if possible. it makes me worry that this code will end up permanent in there and we'll lose our memory of why, and then when something is broken or needing maintenance it will make that work difficult, particularly if we end up removing the referenced package.

alternatively, the worst scenario I see which seems to be quite common is that we have all this special-casing hidden deep in our layers of build scripts and then something breaks and there's no linkage between the failure and the code in the build and testing scripts that leads someone to find the problem.

anyway that's more words than it warrants, but maybe we just ask if we can get rid of `npm run test:php:watch` and not have to deal with this.

Reply:
for now, I'm not saying we can't leave in the exclusion, so don't let that block this PR. but if possible, let's rip it out, as like you said, it shouldn't have been added in the first place

if someone wants watch mode they can still run the scripts manually. they can install the package and use it.

Reply:
@dmsnell Re: the `yoast/phpunit-polyfills` package: your script must be wrong. It is 100% PHP cross-version compatible for PHP 5.4 - 8.2 and provides the PHPUnit cross-version compatibility layer, which is why it allows for a wide range of PHPUnit versions.

Reply:
@jrfnl thanks. I'm sure I'm wrong; I must be wrong. but I like being shown how and why I'm wrong, so if you think the script is wrong, well it's right there and you can point it out.

I'm clearly not an expert on this, but I am looking at the output of `composer show --tree` and for `yoast/phpunit-polyfills` it demonstrably lists dependencies for `phpunit/phpunit` which don't list 5.6 support.

as I noted above, it might make sense to special-case `phpunit` and omit checking its dependencies, _given that we are already assuming the presence of `phpunit`_, but what is wrong in the logic in the script?

I'm not just talking about the dependencies we currently have, but I'm wanting to make sure that when those dependencies change that we don't have to remember that this code exists, that it exists here, and remember to update it properly - not a good track record with relying that much on manual process.

```
yoast/phpunit-polyfills 1.0.4 Set of polyfills for changed PHPUnit functionality to allow for creating PHPUnit cross-version compatible tests
├──php >=5.4
└──phpunit/phpunit ^4.8.36 || ^5.7.21 || ^6.0 || ^7.0 || ^8.0 || ^9.0
   ├──doctrine/instantiator ^1.3.1
   │  └──php ^7.1 || ^8.0
   ├──ext-dom *
   ├──ext-json *
   ├──ext-libxml *
   ├──ext-mbstring *
   │  └──php >=7.1
   ├──ext-xml *
   ├──ext-xmlwriter *
   ├──myclabs/deep-copy ^1.10.1
   │  └──php ^7.1 || ^8.0
```

here's the truncated part of `composer show --tree` showing _some_ of the dependencies not listing PHP 5.6 support. Maybe I'm overlooking that something is optional, or that there's a way to know if these will be supported. I'm definitely relying on package maintainers to list the correct support on their packages.

Reply:
> I'm sure I'm wrong; I must be wrong. but I like being shown how and why I'm wrong, so if you think the script is wrong, well it's right there and you can point it out.

@dmsnell Sorry, I hadn't looked at the script before  - I just responded as the author of the PHPUnit Polyfills.

Looking at the script now, IMO, it is pretty clear where the error is.
1. The script will only work after a `composer install` has been run (as it needs `vendor/autoload`).
2. It will then analyse based on the dependencies which have been installed, which is **highly** dependent on which PHP version was used to run `composer install` (as there is no `composer.lock` file _exactly_ for that reason: to allow for installing the correct PHPUnit version for each PHP version).

So... say I run `composer install` on PHP 8.0, it will install PHPUnit 9.x with the associated dependencies. Neither PHPUnit 9.x, nor any of its dependencies will be compatible with PHP 5.6.

No surprise then that the Polyfills will be flagged as incompatible, but that's not a problem with the polyfills, but a problem with a logic fallacy on which the script is based....

Reply:
> Sorry, I hadn't looked at the script before - I just responded as the author of the PHPUnit Polyfills.

I definitely wish I were smart enough to not have to read the questions I respond to 🙃 

> No surprise then that the Polyfills will be flagged as incompatible, but that's not a problem with the polyfills, but a problem with a logic fallacy on which the script is based....

If I understand what you're saying though it seems to imply the script is correct, the logic in the script is correct, and I simply made the mistake of running it after installing packages with a newer version of PHP, something which wouldn't happen in the CI run.

So I think in a way by trying to show that the script is flawed you actually confirmed it's the right logic, because my goal is to determine programmatically what needs to be removed based on the running script in a way that lets us avoid depending on people manually doing all the right steps.

However…

> highly dependent on which PHP version was used to run composer install…So... say I run composer install on PHP 8.0, it will install PHPUnit 9.x with the associated dependencies. Neither PHPUnit 9.x, nor any of its dependencies will be compatible with PHP 5.6.

When I tried doing this I actually got the same result with PHP 5.6 where the `composer show --tree` output still declares `yoast/phpunit-polyfills` as unsupported, with identical output as listed above.

What I did was:
 - clone the Gutenberg repo in a `docker run php:5.6` environment
 - [install Composer through their script](https://getcomposer.org/download/)
 - run `rm composer.lock && rm -rf vendor && git reset --hard` to make sure that nothing in `vendor` was stale
 - run `./composer.phar install`
 - run `./composer.phar show --list`

```bash
~$ php --version
PHP 5.6.40 (cli) (built: Jan 23 2019 00:04:26)
Copyright (c) 1997-2016 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies
~$ ./composer.phar --version
Composer version 2.2.18 2022-08-20 11:33:38
```

In this case though my script failed to run because of something in `symphony/polyfill` that depended on newer PHP versions. After manually removing `spatie/phpunit-watcher` it progressed further, but still failed to execute the autoloader due to an issue with unsupported `strict_types` inside the `phpunit`'s `Framework/Assets/Functions.php` module.

Maybe my version of Composer is too new? To try and see if this could be the case I tried against with Composer 1.6 from the beginning of 2018 but its output is still the same, still showing `phpunit/phpunit` pulling in dependencies that rely on PHP 7.1 and newer.

So I guess the current state is that if we want to do this programmatically the script is telling us the information from the installed packages as it should, but we need to modify it to read the `Semver` dependencies directly instead of through the autoloader, since that fails on PHP 5.6, and also ensure that we're running it on the right set of installed packages.

```php
require 'phar://composer.phar/vendor/composer/semver/src/Constraint/ConstraintInterface.php';
require 'phar://composer.phar/vendor/composer/semver/src/Constraint/MultiConstraint.php';
require 'phar://composer.phar/vendor/composer/semver/src/Constraint/Constraint.php';
require 'phar://composer.phar/vendor/composer/semver/src/Semver.php';
require 'phar://composer.phar/vendor/composer/semver/src/VersionParser.php';

use Composer\Semver\Semver;

$PHP_VERSION = empty( getenv( 'PHP_VERSION' ) ) ? phpversion() : getenv( 'PHP_VERSION' );
$tree = json_decode( file_get_contents( 'php://stdin' ) );

function works_with_php_version( $version, $package ) {
	foreach ( $package->requires as $dep ) {
		if ( 'php' === $dep->name && ! Semver::satisfies( $version, $dep->version ) ) {
			return false;
		}

		if ( ! empty( $dep->requires ) && false === works_with_php_version( $version, $dep ) ) {
			return false;
		}
	}

	return true;
}

$unsupported = array();
foreach ( $tree->installed as $package ) {
	if ( ! works_with_php_version( $PHP_VERSION, $package ) ) {
		$unsupported[] = $package->name;
	}
}

echo implode( ',', $unsupported );
```

I'd still like to better understand where I'm wrong here because I think I have tried doing what you told me I need to do but the results are the same as when I did it before.

Reply:
> If I understand what you're saying though it seems to imply the script is correct, the logic in the script is correct, and I simply made the mistake of running it after installing packages with a newer version of PHP, something which wouldn't happen in the CI run.

No, as, as you yourself just proved above.
* Your script is intended to find out which dependencies are incompatible so they can be removed **_before_** an install is run.
* Your script requires a Composer install to have been run _already_, which would fail when there are incompatible dependencies.

> Maybe my version of Composer is too new?

For an install on PHP 5.6, you can use the latest of the Composer LTS 2.2.x series.
Composer 2.3.0 and higher require PHP 7.2 (or higher).

> In this case though my script failed to run because of something in symphony/polyfill that depended on newer PHP versions. After manually removing spatie/phpunit-watcher it progressed further, but still failed to execute the autoloader due to an issue with unsupported strict_types inside the phpunit's Framework/Assets/Functions.php module.

I haven't dug in any deeper now, but I suspect this may have something to do with the default PHP version on your system not being PHP 5.6 and something somewhere along the way starting a new PHP process and using the system default PHP version instead of the PHP version you want it to use. Could that be it ?

> So I guess the current state is that if we want to do this programmatically the script is telling us the information from the installed packages as it should, but we need to modify it to read the Semver dependencies directly instead of through the autoloader, since that fails on PHP 5.6, and also ensure that we're running it on the right set of installed packages.

Possibly, and while I appreciate you are enjoying the intellectual exercise of trying to understand this. @anton-vlasenko 's two liner in the GHA script fixes this with a lot less effort and higher accuracy.

Reply:
> Your script is intended to find out which dependencies are incompatible so they can be removed before an install is run.

that's not the case, but that's how I interpreted your statement when you said "which is highly dependent on which PHP version was used to run composer install". if the running PHP version determines what packages composer will install then I took that to imply it wouldn't install packages that are listed as incompatible.

> For an install on PHP 5.6, you can use the latest of the Composer LTS 2.2.x series.

thanks. as indicated in my previous comment I did attempt this on 2.2 and 1.6 versions.

> system default PHP version instead of the PHP version you want it to use. Could that be it ?

No I don't think so, because as indicated in my previous comment, PHP 5.6 was the only version of PHP installed within the Docker container. It was both the system default and the only available version to run.